silkworm/db/snapshots/index/btree.hpp

```diff
@@ -0,0 +1,104 @@
+/*
+   Copyright 2024 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#pragma once
+
+#include <functional>
+#include <tuple>
+#include <vector>
+
+#include <silkworm/core/common/bytes.hpp>
+
+#include "../seg/decompressor.hpp"
+
+namespace silkworm::snapshots::index {
+
+class BTree {
+  public:
+    using DataIterator = seg::Decompressor::Iterator;
+    using DataIndex = uint64_t;
+    using KeyValue = std::pair<Bytes, Bytes>;
+    using LookupResult = std::pair<bool, KeyValue>;
+    using DataLookup = std::function<LookupResult(DataIndex, DataIterator&)>;
+    using CompareResult = std::pair<int, Bytes>;
+    using KeyCompare = std::function<CompareResult(ByteView, DataIndex, DataIterator&)>;
+
+    using SeekResult = std::tuple<bool, Bytes, Bytes, DataIndex>;
+    using GetResult = std::tuple<bool, Bytes, DataIndex>;
+
+    BTree(uint64_t num_nodes,
+          uint64_t fanout,
+          DataLookup data_lookup,
+          KeyCompare compare_key,
+          DataIterator& data_it,
+          std::span<uint8_t> encoded_nodes = {});
+
+    //! \brief Search and return first key-value pair w/ key greater than or equal to \p seek_key
+    //! \param seek_key the key to look for
+    //! \param data_it an iterator to the key-value data sequence
+    //! \return tuple (found, key, value, data index)
+    //! \verbatim
+    //! - found is true iff exact key match is encountered
+    //! - if seek_key is empty, return first key and found=true
+    //! - if found item.key has \p seek_key as prefix, return found=false and item.key
+    //! - if key is greater than all keys, return found=false and empty key
+    //! \endverbatim
+    SeekResult seek(ByteView seek_key, DataIterator& data_it);
```

Comment:
No, it is not because it's not a pure output parameter: it's an input parameter that gets modified by the function call (i.e. the iterator gets moved).